Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,270 @@
+		return SignalProducer<URL, CarthageError>(buildProductURLs)
+		.flatMap(.merge) { url -> SignalProducer<(), CarthageError> in
+			let platformName = url.deletingLastPathComponent().lastPathComponent
+			let frameworkName = url.deletingPathExtension().lastPathComponent
+			let frameworkURL = url.appendingPathComponent(frameworkName, isDirectory: false)
+			return md5ForFileAtURL(frameworkURL)
+				.attemptMap { md5 -> Result<(), CarthageError> in
+					let cachedFramework = CachedFramework(name: frameworkName, md5: md5)
+					if var frameworks = platformCaches[platformName] {
+						frameworks.append(cachedFramework)
+						platformCaches[platformName] = frameworks
+						return .success(())
+					}
+					else {
+						return .failure(.versionFileError(description: "unexpected platform found in path"))
+					}
+				}
+				.collect()
+				.flatMap(.concat) { _ -> SignalProducer<(), CarthageError> in
+					return writeVersionFile
+				}
+		}
+	} else {
+		// Write out an empty version file for dependencies with no built frameworks, so cache builds can differentiate between
+		// no cache and a dependency that has no frameworks
+		return writeVersionFile
+	}
+}
+
+/// Determines whether a dependency can be skipped because it is
+/// already cached.
+///
+/// If a set of platforms is not provided and a version file exists,
+/// the platforms in the version file are used instead.
+///
+/// Returns true if the the dependency can be skipped.
+public func versionFileMatchesDependency(_ dependency: Dependency<PinnedVersion>, forPlatforms platforms: Set<Platform>, rootDirectoryURL: URL) -> SignalProducer<Bool, CarthageError> {
+	let rootBinariesURL = rootDirectoryURL.appendingPathComponent(CarthageBinariesFolderPath, isDirectory: true).resolvingSymlinksInPath()
+	let versionFileURL = rootBinariesURL.appendingPathComponent(".\(dependency.project.name).version")
+	guard let versionFile = VersionFile(url: versionFileURL) else {
+		return .init(value: false)
+	}
+	let commitish = dependency.version.commitish
+	
+	let cachedPlatforms = versionFile.cachedPlatforms()
+	let platformsToCheck = platforms.isEmpty ? cachedPlatforms : platforms
+
+	return currentXcodeVersion()
+		.flatMap(.concat) { xcodeVersion in
+			return SignalProducer<Platform, CarthageError>(platformsToCheck)
+				.flatMap(.merge) { platform in
+					return versionFile.check(platform, commitish: commitish, xcodeVersion: xcodeVersion, rootDirectoryURL: rootDirectoryURL)
+				}
+				.reduce(true) { current, result in
+					return current && result
+				}
+		}
+}
+
+private func currentXcodeVersion() -> SignalProducer<String, CarthageError> {
+	let task = Task("/usr/bin/xcrun", arguments: ["xcodebuild", "-version"])
+	return task.launch()
+		.mapError(CarthageError.taskError)
+		.ignoreTaskData()
+		.attemptMap { data in
+			guard let versionString = String(data: data, encoding: .utf8) else {
+				return .failure(.versionFileError(description: "Could not get xcode version"))
```

Comment:
Ok, looked around, and it appears 'parseError' is used in other similar situations (like when a git task fails to read from ls-tree), so it's just the comment that's wrong. Using parse error here as well, and updating the comment